src/API.jl

```diff
@@ -1208,13 +1208,35 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
 
     # find and guard against circular deps
     circular_deps = Base.PkgId[]
-    function in_deps(_pkgs, deps, dmap)
-        isempty(deps) && return false
-        !isempty(intersect(_pkgs, deps)) && return true
-        return any(dep->in_deps(vcat(_pkgs, dep), dmap[dep], dmap), deps)
+    # Three states
+    # !haskey -> never visited
+    # true -> could be part of a cycle
+    # false -> not part of a cycle
+    could_be_cycle = Dict{Base.PkgId, Bool}()
+    dfs_num = 0
+    function scan_pkg!(pkg, dmap)
+        did_visit_dep = true
+        inpath = get!(could_be_cycle, pkg) do
+            did_visit_dep = false
+            return true
+        end
+        if did_visit_dep ? inpath : scan_deps!(pkg, dmap)
+            # Found a cycle. Delete this and all parents
+            # to let this cycle be found again by another toplevel dep
+            delete!(could_be_cycle, pkg)
+            return true
+        end
+        return false
```

Comment:
no reason to visit this again later, since it is proven here to be not in a cycle
```suggestion
        could_be_cycle[pkg] = false
        return false
```

Reply:
I don't see why this is necessarily better here

Reply:
It just does a useless extra dict lookup in the hot path.

Reply:
Ah, right they are equivalent but I misread it and thought all of the dict manipulation needed to be from the same function--it doesn't.